src/devices/bus/hp80_io/82937.cpp

```diff
@@ -0,0 +1,354 @@
+		}
+	} else {
+		if (m_ieee488->eoi_r()) {
+			BIT_SET(res , P1_EOI_BIT);
+		}
+		if (m_ieee488->dav_r()) {
+			BIT_SET(res , P1_DAV_BIT);
+		}
+		if (ndac) {
+			BIT_SET(res , P1_NDAC_BIT);
+		}
+		BIT_SET(res , P1_NRFD_BIT);
+	}
+
+	return res;
+}
+
+WRITE8_MEMBER(hp82937_io_card_device::p1_w)
+{
+	update_signals();
+	update_data_out();
+}
+
+READ8_MEMBER(hp82937_io_card_device::dio_r)
+{
+	if (m_dio_out) {
+		return 0xff;
+	} else {
+		return m_ieee488->dio_r();
+	}
+}
+
+WRITE8_MEMBER(hp82937_io_card_device::dio_w)
+{
+	update_data_out();
+}
+
+READ8_MEMBER(hp82937_io_card_device::switch_r)
+{
+	return m_sw1->read() | 0xc0;
+}
+
+WRITE8_MEMBER(hp82937_io_card_device::latch_w)
+{
+	LOG("latch=%02x\n" , data);
+	m_latch = data;
+	update_signals();
+	update_data_out();
+}
+
+WRITE_LINE_MEMBER(hp82937_io_card_device::ieee488_ctrl_w)
+{
+	update_signals();
+	update_data_out();
+}
+
+static INPUT_PORTS_START(hp82937_port)
+	MCFG_HP80_IO_SC(7)
+	PORT_START("sw1")
+	PORT_DIPNAME(0x1f , 0x15 , "HPIB address")
+	PORT_DIPLOCATION("S1:7,6,5,4,3")
+	PORT_DIPSETTING(0x00 , "0")
+	PORT_DIPSETTING(0x01 , "1")
+	PORT_DIPSETTING(0x02 , "2")
+	PORT_DIPSETTING(0x03 , "3")
+	PORT_DIPSETTING(0x04 , "4")
+	PORT_DIPSETTING(0x05 , "5")
+	PORT_DIPSETTING(0x06 , "6")
+	PORT_DIPSETTING(0x07 , "7")
+	PORT_DIPSETTING(0x08 , "8")
+	PORT_DIPSETTING(0x09 , "9")
+	PORT_DIPSETTING(0x0a , "10")
+	PORT_DIPSETTING(0x0b , "11")
+	PORT_DIPSETTING(0x0c , "12")
+	PORT_DIPSETTING(0x0d , "13")
+	PORT_DIPSETTING(0x0e , "14")
+	PORT_DIPSETTING(0x0f , "15")
+	PORT_DIPSETTING(0x10 , "16")
+	PORT_DIPSETTING(0x11 , "17")
+	PORT_DIPSETTING(0x12 , "18")
+	PORT_DIPSETTING(0x13 , "19")
+	PORT_DIPSETTING(0x14 , "20")
+	PORT_DIPSETTING(0x15 , "21")
+	PORT_DIPSETTING(0x16 , "22")
+	PORT_DIPSETTING(0x17 , "23")
+	PORT_DIPSETTING(0x18 , "24")
+	PORT_DIPSETTING(0x19 , "25")
+	PORT_DIPSETTING(0x1a , "26")
+	PORT_DIPSETTING(0x1b , "27")
+	PORT_DIPSETTING(0x1c , "28")
+	PORT_DIPSETTING(0x1d , "29")
+	PORT_DIPSETTING(0x1e , "30")
+	PORT_DIPSETTING(0x1f , "31")
+	PORT_DIPNAME(0x20 , 0x20 , "Sys. controller")
+	PORT_DIPLOCATION("S1:2")
+	PORT_DIPSETTING(0x00 , DEF_STR(Off))
+	PORT_DIPSETTING(0x20 , DEF_STR(On))
+INPUT_PORTS_END
+
+ioport_constructor hp82937_io_card_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(hp82937_port);
+}
+
+void hp82937_io_card_device::device_start()
+{
+	save_item(NAME(m_dio_out));
+	save_item(NAME(m_talker_out));
+	save_item(NAME(m_iatn));
+	save_item(NAME(m_latch));
+}
+
+void hp82937_io_card_device::device_reset()
+{
+	hp80_io_card_device::device_reset();
+
+	m_latch = 0;
+	m_updating = false;
+	update_signals();
+	update_data_out();
+}
+
+void hp82937_io_card_device::update_data_out()
+{
+	m_ieee488->dio_w(m_dio_out ? m_cpu->p2_r(machine().dummy_space() , 0) : 0xff);
+}
+
+void hp82937_io_card_device::update_signals()
+{
+	// Avoid recursive re-enter when writing to IEEE488 signals
+	if (m_updating) {
+		return;
+	}
+	m_updating = true;
```

Comment:
This pattern for avoiding mutual recursion is usually a symptom of a less-than-ideal design.  Is there a way to rethink the bus to avoid the need to use this pattern?

(Also, for this to be exception-safe, you need to use a local object destructor or a catch-all block to ensure `m_updating` will be cleared.  Doing it this way can lead to a stuck flag.)